omegaconf/nodes.py

```diff
@@ -33,15 +35,19 @@ def _set_value(self, value: Any, flags: Optional[Dict[str, bool]] = None) -> Non
         if self._get_flag("readonly"):
             raise ReadonlyConfigError("Cannot set value of read-only config node")
 
-        if isinstance(value, str) and get_value_kind(value) in (
-            ValueKind.INTERPOLATION,
-            ValueKind.MANDATORY_MISSING,
-        ):
-            self._val = value
-        else:
+        val = None
+        if isinstance(value, str):
+            vk = get_value_kind(value)
+            if vk == ValueKind.INTERPOLATION:
+                grammar_parser.parse(value)  # validate syntax
+                val = value
+            elif vk == ValueKind.MANDATORY_MISSING:
+                val = value
+        if val is None:
```

Comment:
Pushing `grammar_parser.parse(value)` into get_value_kind will also allow you to simplify this back.

Reply:
> Pushing `grammar_parser.parse(value)` into get_value_kind will also allow you to simplify this back.

Or you could just drop support for Python <3.8 so I can use `:=` :p